langkit/compiled_types.py

```diff
@@ -3626,6 +3653,113 @@ def require_hash_function(self):
         self.element_type.require_hash_function()
 
 
+class IteratorType(CompiledType):
+    """
+    Base class for iterator types.
+    """
+    def __init__(self, name, element_type):
+        # For now, we expose all iterator types since their only purpose is
+        # to be used in the different public APIs and they don't even have a
+        # DSL API.
```

Comment:
This is valid only if the `element_type` can be exposed itself, and we don’t know that at this point. I suggest to do as for arrays: consider it’s not exposed at construction, and enhance the `expose_public_api_types` pass to enable public API exposition when needed.

Besides, our ultimate goal is to use iterators internally, so I’m not sure we want to set in store that iterators are only for exposed types. If we do this nonetheless, please add a testcase that shows that we reject iterators on, say, lexical envs (which can’t be exposed).